[pdata] Add documentation defining naming conventions (#6255)
open-telemetry/opentelemetry-collector: pdata/README.md

> # Pipeline data (pdata)

Reviewer: Add one section about the "flags". I think we had a good conversation with conclusions there as well.

Author: Added sections for enums and flags

Pipeline data (pdata) implements data structures that represent telemetry data in-memory. All data received 
is converted into this format, travels through the pipeline in this format, and is converted from this format by 
exporters when sending.

Current implementation primarily uses OTLP protobuf structs as the underlying data structures for many of the 
declared structs. We keep a pointer to OTLP protobuf in the "orig" member field. This allows efficient translation 
to/from OTLP wire protocol. The underlying data structure is kept private so that we are free to make changes to it 
in the future.

API provided by pdata ensures that no underlying data can be used in more than one place. Each pdata instance cannot 
contain a reference to an object that is already used somewhere else.

## API naming convention

> ### protobuf message representation in pdata

Reviewer: One more exception is that we have "Resource" in the common package instead of "resource" package.

Author: We don't specify package names here. I added a small paragraph for that

Pipeline data structs SHOULD be based on the names of the underlying OTLP protobuf messages. Data types for 
protobuf messages defined as part of another message SHOULD include the owner's name as a prefix. Exceptions to these 
rules are possible, but not encouraged. Another name can be chosen for brevity or if the struct provides a 
different interface to manipulate the underlying data. The following exceptions are currently accepted:

- `plog.Logs` based on `LogsData` protobuf message.
- `pmetric.Metrics` based on `MetricsData` protobuf message.
- `ptrace.Traces` based on `TracesData` protobuf message.
- `pcommon.Slice` based on `ArrayValue` protobuf message.
- `pcommon.Map` based on `KeyValueList` protobuf message.
- `pcommon.Value` based on `AnyValue` protobuf message.

Each pdata struct MUST have an initialization function starting with `New` prefix. Usage of zero-initialized is 

> prohibited and can case panic. Each pdata struct MUST provide the following methods:

Reviewer 2: ```suggestion
prohibited and can cause a panic. Each pdata struct MUST provide the following methods:
```

- `MoveTo(<type>)`: moves all the data from one struct instance to another. Destination instance is overridden, the 

>   source instance is re-initialized.

Reviewer: Clarify what "re-initialized" mean.

Author: done

- `CopyTo(<type>)`: copies all the data from one struct instance to another. Destination instance is overridden, the
  source instance is not modified.

### protobuf fields representation in pdata

### Singular fields

Each pdata struct based on a protobuf message SHOULD have getter methods for every protobuf field called the same as

> the protobuf field but in CamelCase. Exceptions to this rule are allowed if exposing a field is not desirable in pdata. 

Author: Added couple of examples. I don't think we should necessary limit reasons why fields should/can be not exposed in pdata. Just wanted to keep it open.

pdata can also provide a different interface to manipulate fields without exposing them explicitly as it's done in 
`pcommon.Map`. If protobuf field is of scalar or enum type, the corresponding pdata struct MUST provide a setter 
method with `Set` prefix.

Some fields of scalar type in protobuf message MAY be represented by another pdata type providing an additional 
interface, for example `pcommon.Timestamp`.

### Optional fields

Each optional field in protobuf message exposed in pdata MUST have an additional method with `Has` prefix to
determine if the field is set.

### OneOf fields

> A pdata struct representing a protobuf message with an oneof field MUST provide getter and setter methods for each 
> option of the oneof field. If an oneof field option is another protobuf message, setter name MUST include `Empty` 

Reviewer 2: 'oneof' starts with a consonant sound /w/ so it should be preceded by `a`, not `an` (see [here](https://english.stackexchange.com/a/333972))
```suggestion
A pdata struct representing a protobuf message with a oneof field MUST provide getter and setter methods for each 
option of the oneof field. If a oneof field option is another protobuf message, setter name MUST include `Empty` 
```

# Pipeline data (pdata)

Pipeline data (pdata) implements data structures that represent telemetry data in-memory. All data received 
is converted into this format, travels through the pipeline in this format, and is converted from this format by 
exporters when sending.

Current implementation primarily uses OTLP protobuf structs as the underlying data structures for many of the 
declared structs. We keep a pointer to OTLP protobuf in the "orig" member field. This allows efficient translation 
to/from OTLP wire protocol. The underlying data structure is kept private so that we are free to make changes to it 
in the future.

> API provided by pdata ensures that no underlying data can be used in more than one place. Each pdata instance cannot 

Reviewer 3: >no underlying data can be used in more than one place

This is a bit unclear, since "used" and "place" can be interpreted in different ways.

Can we maybe say instead:
```suggestion
The pdata is designed to avoid mutable data sharing and bugs that stem from that. Each pdata instance cannot 
```

Author: Updated

contain a reference to an object that is already used somewhere else.

## API naming convention

### Protobuf message representation in pdata

Packages of pdata are not following names of the protobuf packages. pdata has a package per telemetry type starting 

> with `p`, e.g. `ptace`, and `pcommon` package with API for other protobuf definitions from `common` and `resource` 

Reviewer 3: ```suggestion
with `p`, e.g. `ptrace`, and `pcommon` package with API for other protobuf definitions from `common` and `resource` 
```

Reviewer 3: >with API for other protobuf definitions from `common` and `resource`
>protobuf packages

I do not quite understand what this means.

Author: Updated. Thanks!

protobuf packages.

Pipeline data structs SHOULD be based on the names of the underlying OTLP protobuf messages. Data types for 

> protobuf messages defined as part of another message SHOULD include the owner's name as a prefix. Exceptions to these 

Reviewer 3: Can we add one example to make clear what this means?

Author: Added

rules are possible, but not encouraged. Another name can be chosen for brevity or if the struct provides a completely
different interface to manipulate the underlying data. The following exceptions are currently accepted:

- `plog.Logs` based on `LogsData` protobuf message.
- `pmetric.Metrics` based on `MetricsData` protobuf message.
- `ptrace.Traces` based on `TracesData` protobuf message.
- `pcommon.Slice` based on `ArrayValue` protobuf message.
- `pcommon.Map` based on `KeyValueList` protobuf message.

> - `pcommon.Value` based on `AnyValue` protobuf message.

Reviewer 3: Would it be better to delete all these lines and only keep one exception example, to reduce the amount of maintenance this document requires if the source code is refactored? Let the source code be the source of truth. If you want the fact of the exception to be documented it would be better to add a comment in the source code near the name.

Author: Makes sense. Removed

Each pdata struct MUST have an initialization function starting with `New` prefix. Usage of zero-initialized values is 
prohibited and can cause a panic. Each pdata struct MUST provide the following methods:

- `MoveTo(<type>)`: moves all the data from one struct instance to another. The destination instance is overwritten and the 
  source instance is re-initialized as a new empty instance.

> - `CopyTo(<type>)`: copies all the data from one struct instance to another. The destination instance is overwritten and the 

Reviewer 3: ```suggestion
- `CopyTo(<type>)`: deep copies all the data from one struct instance to another. The destination instance is overwritten and the 
```

Author: applied

   include name of the oneof field along with name of the option. For example:

```go
func (ms NumberDataPoint) IntValue() int64
func (ms NumberDataPoint) SetIntValue(v int64)
```

Additionally, the pdata struct MUST provide a type getter for every exposed oneof protobuf field. Name of the getter 
MUST be `Type` for oneof fields representing the whole protobuf message (1), or `<oneof field>Type` for oneof fields 
relevant only to a particular field. The function MUST return an `int32` enum type called `<struct name>Type` for (1) 
and `<struct name><oneof field>Type` for (2). For example:

- `func (ms Metric) Type() MetricType` (1)
- `func (ms NumberDataPoint) ValueType() NumberDataPointValueType` (2)

The enum constants MUST be called the same as the type with a suffix named after the oneof option, e.g.: 

- `MetricTypeSum` (1)
- `NumberDataPointValueTypeString` (2)

An unset oneof protobuf value MUST be represented by a pdata constant with `Empty` suffix, e.g.:

- `MetricTypeEmpty` (1)
- `NumberDataPointValueTypeEmpty` (2)

The pdata enum type SHOULD have `String()` method returning a string value of the corresponding oneof field option. 

> ### Repeated fields

Reviewer 4: ```suggestion
### Repeated protobuf fields
```

- `Sort(func(a, b <element type>) bool) <slice type>`: sorts the slice using the function and returns itself.

### Repeated scalar fields

Repeated scalar protobuf fields MUST be represented as another pdata type wrapping native go slices. Name of the 
type MUST include name of the primitive data type ending with `Slice`, e.g. `UInt64Slice`. Each pdata type MUST have 
the following methods:

- `AsRaw() []<primitive type>`: returns copy of the underlying go slice. 
- `FromRaw(val []<primitive type>)`: copies a go slice into the pdata wrapper instance.
- `Len()`: returns length of the slice.
- `At(int) <primitive type>`: returns an element from the slice by its index, can cause "index out of range" panic.
- `EnsureCapacity(int)`: ensures that the slice have at least the requested capacity.
- `AppendEmpty() <primitive type>`: adds another slice element with a zero value.
- `CopyTo(<slice type>)`: copies the slice to another slice of the same type overriding the destination.
- `MoveTo(<slice type>)`: move the slice to another slice of the same type. The destination is overridden, the 

>   source is set to an empty slice.

Reviewer 3: Do we need these detailed docs given that this is all codified in the pdata generator? I am worried that we may forget to keep this up to date when we change any behavior in the generator.
The list of methods also seems incomplete, I can't find `MoveAndAppendTo` or `RemoveIf`.

Author: Makes sense. Maybe requiring a particular set of methods is redundant here. Removed them

- If a protobuf enum is defined on the package level, pdata type SHOULD have the same name.
- If a protobuf enum is defined as part of another message, pdata type SHOULD include the protobuf message name as a 
  prefix.

Constants defined for the pdata int64 type MUST have the same names and numeric values defined in protobuf. Names 
of the constants must be translated from ALL_CAPS_SNAKE_CASE to CamelCase.

The pdata enum type SHOULD have `String()` method returning a string value for each constant.

Example of a protobuf enum definition in pdata:

```golang
type AggregationTemporality int32
const (
	AggregationTemporalityUnspecified = AggregationTemporality(otlpmetrics.AggregationTemporality_AGGREGATION_TEMPORALITY_UNSPECIFIED)
	AggregationTemporalityDelta = AggregationTemporality(otlpmetrics.AggregationTemporality_AGGREGATION_TEMPORALITY_DELTA)
	AggregationTemporalityCumulative = AggregationTemporality(otlpmetrics.AggregationTemporality_AGGREGATION_TEMPORALITY_CUMULATIVE)
)
```

### Flags

Flags fields are typically defined in protobuf as `uint32` fields representing 32 distinct boolean flags. The bit 

> fields of the flags are defined as enum definitions. The exposed protobuf flags fields MUST be defined as new types 

Author: Ok, removed the sentence.

> with `uint32` underlying type and called according to the same rules as struct names representing protobuf messages 

Reviewer: Mention that is immutable.

Author: done

and other enum types. Each pdata flags type MUST have an empty variable of the fields type with `Default` prefix. 

> Each flag MUST have a getter method returning a boolean value and a setter method with `With` prefix returning a 

Reviewer: Not necessary boolean, because we may have non boolean fields in the flags (a 3 state for example).

Author: corrected
